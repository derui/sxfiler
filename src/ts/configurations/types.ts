import { ObjectEnum } from "@/utils";

export type ItemKey<T> = readonly [string, string, string];
export type SectionKey = readonly [string, string];
export type CategoryKey = readonly [string];

export type Descriptor<T> = {
  readonly key: T;
  readonly displayName: string;
  readonly description: string;
};

export type CategoryDescriptor = Descriptor<CategoryKey>;

export type SectionDescriptor = Descriptor<SectionKey>;

export type ItemDescriptor = Descriptor<ItemKey<any>>;

export type SelectOption = {
  value: any;
  toString(): string;
  display: string;
};

export const ItemElementType = {
  TEXT: "TEXT",
  NUMBER: "NUMBER",
  BOOLEAN: "BOOLEAN",
  SELECT_ONE: "SELECT_ONE",
} as const;
export type ItemElementType = ObjectEnum<typeof ItemElementType>;

export type ItemKind<K extends ItemElementType> = {
  kind: K;
};

export type TextItemType = ItemKind<"TEXT"> & {
  defaultValue?: string;
};
export type NumberItemType = ItemKind<"NUMBER"> & {
  defaultValue?: number;
};
export type BooleanItemType = ItemKind<"BOOLEAN"> & {
  defaultValue: boolean;
};
export type SelectOneItemType = ItemKind<"SELECT_ONE"> & {
  options: SelectOption[];
  defaultValue: string;
};

export type ItemType = TextItemType | NumberItemType | BooleanItemType | SelectOneItemType;

export type Item = ItemDescriptor & {
  readonly type: ItemType;
};

export type Section = SectionDescriptor & {
  readonly items: Item[];
};

export type Category = CategoryDescriptor & {
  readonly sections: Section[];
};
