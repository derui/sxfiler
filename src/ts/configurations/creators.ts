import { Section, Category, Item, ItemType, ItemKey, SectionDescriptor, CategoryDescriptor } from "./types";

export const createCategory = function (args: CategoryDescriptor & { sections: Section[] }): Category {
  return { ...args };
};

export const createSection = function (args: SectionDescriptor & { items: Item[] }): Section {
  return { ...args };
};

export const createItem = function (args: {
  key: ItemKey<any>;
  displayName: string;
  description: string;
  type: ItemType;
}): Item {
  const { key, ...rest } = args;
  return {
    key,
    displayName: rest.displayName,
    description: rest.description,
    type: rest.type,
  };
};
