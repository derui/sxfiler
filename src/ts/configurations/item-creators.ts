import { ItemElementType, ItemType, SelectOption } from "./types";

export const createText = (value?: string): ItemType => ({ kind: ItemElementType.TEXT, defaultValue: value });

export const createNumber = (value?: number): ItemType => ({ kind: ItemElementType.NUMBER, defaultValue: value });

export const createBoolean = (value: boolean): ItemType => ({ kind: ItemElementType.BOOLEAN, defaultValue: value });

export const createSelectOne = (options: SelectOption[], value: string): ItemType => ({
  kind: ItemElementType.SELECT_ONE,
  options,
  defaultValue: value,
});
