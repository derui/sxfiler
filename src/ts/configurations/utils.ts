import { ItemElementType, Item } from "./types";

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const isTypedText = function (arg: Item, value: any): value is string {
  return arg.type.kind === ItemElementType.TEXT && value !== undefined && value !== null;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const isTypedNumber = function (arg: Item, value: any): value is number {
  return arg.type.kind === ItemElementType.NUMBER && value !== undefined && value !== null;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const isTypedBoolean = function (arg: Item, value: any): value is boolean {
  return arg.type.kind === ItemElementType.BOOLEAN && value !== undefined && value !== null;
};
