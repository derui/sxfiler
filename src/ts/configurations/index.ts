import { SortType, SortTypeMap } from "@/generated/types_pb";
import { Category, Item, Section, ItemKey, SectionKey, CategoryKey } from "./types";
import { createCategory, createSection, createItem } from "./creators";
import * as ItemCreators from "./item-creators";

const itemKeyFrom = <T>(section: SectionKey, key: string): ItemKey<T> => [section[0], section[1], key];
const categoryKeyOf = (key: ItemKey<any>): CategoryKey => [key[0]];
const sectionKeyOf = (key: ItemKey<any>): SectionKey => [key[0], key[1]];
const sectionKeyFrom = (category: CategoryKey, key: string): SectionKey => [category[0], key];
export const qualified = (key: readonly string[]): string => key.join(".");

export const findItemBy = (key: ItemKey<any>, category: Category): Item | undefined => {
  const categoryKey = categoryKeyOf(key);
  const sectionKey = sectionKeyOf(key);
  if (qualified(category.key) !== qualified(categoryKey)) {
    return;
  }
  const section = category.sections.find((v) => v.key === sectionKey);
  if (!section) {
    return;
  }

  return section.items.find((v) => qualified(v.key) === qualified(key));
};

export const findSectionBy = (key: SectionKey, category: Category): Section | undefined => {
  return category.sections.find((v) => qualified(v.key) === qualified(key));
};

export const findValueBy = (configuration: { [p: string]: any }, item: Item): any => {
  const value = configuration[qualified(item.key)];
  if (value === undefined || value === null) {
    return;
  }

  return value;
};

export const categories = {
  general: ["general"],
} as const;

export const sections = {
  behaviors: sectionKeyFrom(categories.general, "behaviors"),
  theme: sectionKeyFrom(categories.general, "theme"),
} as const;

export const itemKeys = {
  general: {
    behaviors: {
      confirmationWhenDelete: itemKeyFrom<string>(sections.behaviors, "confirmation_when_delete"),
      maxNumberOfHistories: itemKeyFrom<number>(sections.behaviors, "max_number_of_histories"),
      defaultSortOrder: itemKeyFrom<SortTypeMap>(sections.behaviors, "default_sort_order"),
    },
    theme: {
      currentTheme: itemKeyFrom<string>(sections.theme, "current_theme"),
    },
  },
} as const;

export namespace Configurations {
  const general = createCategory({
    key: categories.general,
    displayName: "General",
    description: "General configurations",
    sections: [
      createSection({
        key: sections.behaviors,
        displayName: "Behaviors",
        description: "Behaviors",
        items: [
          createItem({
            key: itemKeys.general.behaviors.confirmationWhenDelete,
            displayName: "Confirmation when delete",
            description: "Need user confirmation when delete something",
            type: ItemCreators.createBoolean(true),
          }),
          createItem({
            key: itemKeys.general.behaviors.maxNumberOfHistories,
            displayName: "Maximum number of histories",
            description:
              "History is the history of moved directory in a pane. This setting defines maximum number of histories in a pane",
            type: ItemCreators.createNumber(100),
          }),
          createItem({
            key: itemKeys.general.behaviors.defaultSortOrder,
            displayName: "Default sort order of Pane",
            description: "This setting defines sort order of filer when open this application",
            type: ItemCreators.createSelectOne(
              [
                {
                  display: "Date",
                  toString() {
                    return `${SortType.DATE}`;
                  },
                  value: SortType.DATE,
                },
                {
                  display: "Name",
                  toString() {
                    return `${SortType.NAME}`;
                  },
                  value: SortType.NAME,
                },
                {
                  display: "Size",
                  toString() {
                    return `${SortType.SIZE}`;
                  },
                  value: SortType.SIZE,
                },
              ],
              `${SortType.NAME}`
            ),
          }),
        ],
      }),
      createSection({
        key: sections.theme,
        displayName: "Theme",
        description: "Theme",
        items: [
          createItem({
            key: itemKeys.general.theme.currentTheme,
            displayName: "Current theme",
            description: "Select current theme",
            type: ItemCreators.createText("default"),
          }),
        ],
      }),
    ],
  });

  export const definitions = {
    general,
  } as const;
}
