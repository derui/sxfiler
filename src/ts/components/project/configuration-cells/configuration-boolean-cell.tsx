import { h } from "preact";
import { useState, StateUpdater } from "preact/hooks";
import { Component as Switch } from "@/components/ui/switch";
import { Item, ItemElementType } from "@/configurations/types";
import { isTypedBoolean } from "@/configurations/utils";

type Callback = (value: boolean) => void;

export type Props = {
  item: Item;
  value?: boolean;
  onUpdated: Callback;
};

const handleChange = (callback: Callback, stateUpdater: StateUpdater<boolean>) => (v: boolean) => {
  stateUpdater(v);
  callback(v);
};

export const Component: preact.FunctionComponent<Props> = ({ item, onUpdated, value }) => {
  if (item.type.kind !== ItemElementType.BOOLEAN) {
    return null;
  }
  let v = item.type.defaultValue;
  if (isTypedBoolean(item, value)) {
    v = value;
  }

  const [state, updateState] = useState(v);

  return (
    <div class="configuration-cell__root" data-testid="configuration-boolean-cell-root">
      <label class="configuration-cell__label">{item.displayName}</label>
      <div class="configuration-cell__switch-container" data-testid="configuration-boolean-cell-container">
        <Switch checked={state} onChange={handleChange(onUpdated, updateState)} />
      </div>
      <p class="configuration-cell__description">{item.description}</p>
    </div>
  );
};
