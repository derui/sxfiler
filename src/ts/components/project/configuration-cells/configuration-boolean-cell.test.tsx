import { h } from "preact";

import { render, cleanup, fireEvent, act } from "@testing-library/preact";
import { render as renderAsText } from "preact-render-to-string";
import { Component as Cell } from "./configuration-boolean-cell";
import { createItem, toItemKey } from "@/configurations/creators";
import { createBoolean, createNumber } from "@/configurations/item-creators";

describe("Project", () => {
  describe("Configuration Boolean Cell", () => {
    afterEach(() => {
      cleanup();
    });

    it("render correctly", () => {
      const item = createItem({
        key: toItemKey(["foo", "bar", "baz"]),
        displayName: "name",
        description: "description",
        type: createBoolean(false),
      });
      const tree = renderAsText(<Cell item={item} value={true} onUpdated={() => {}} />);

      expect(tree).toMatchSnapshot();
    });

    it("invoke callback every event", async () => {
      const item = createItem({
        key: toItemKey(["foo", "bar", "baz"]),
        displayName: "name",
        description: "description",
        type: createBoolean(false),
      });

      const callback = jest.fn();
      const query = render(<Cell item={item} onUpdated={callback} />);
      const input = await query.findByTestId("switch-root");

      act(() => {
        fireEvent.click(input);
      });

      expect(callback).toBeCalledWith(true);
    });

    it("do not render component if definition is not boolean type", () => {
      const item = createItem({
        key: toItemKey(["foo", "bar", "baz"]),
        displayName: "name",
        description: "description",
        type: createNumber(10),
      });

      const callback = jest.fn();
      const query = render(<Cell item={item} onUpdated={callback} />);
      const input = query.queryByTestId("configuration-boolean-cell-container");

      expect(input).toBeNull();
    });
  });
});
