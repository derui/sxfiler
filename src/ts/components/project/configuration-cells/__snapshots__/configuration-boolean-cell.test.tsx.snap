// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Configuration Boolean Cell render correctly 1`] = `
"<div class=\\"configuration-cell__root\\" data-testid=\\"configuration-boolean-cell-root\\"><label class=\\"configuration-cell__label\\">name</label><div class=\\"configuration-cell__switch-container\\" data-testid=\\"configuration-boolean-cell-container\\"><span class=\\"switch\\" data-testid=\\"switch-root\\"><span class=\\"
switch__rail
switch__rail--checked
\\"><span class=\\"
switch__box
switch__box--checked
\\"></span></span><input class=\\"switch__input\\" type=\\"checkbox\\" checked /></span></div><p class=\\"configuration-cell__description\\">description</p></div>"
`;
